include/range/v3/view/join.hpp

```diff
@@ -58,259 +65,253 @@ namespace ranges
         // Join a range of ranges
         template<typename Rng>
         struct join_view<Rng, void>
-          : view_adaptor<join_view<Rng, void>, Rng,
-                detail::join_cardinality<
-                    range_cardinality<Rng>,
-                    range_cardinality<range_value_t<Rng>>>::value>
+          : view_facade<join_view<Rng, void>, detail::join_cardinality<Rng>::value>
         {
-        private:
-            CONCEPT_ASSERT(Range<Rng>());
-            CONCEPT_ASSERT(Range<range_reference_t<Rng>>());
-            using size_t_ = common_type_t<range_size_t<Rng>, range_size_t<range_value_t<Rng>>>;
+            using size_type = common_type_t<range_size_t<Rng>, range_size_t<range_reference_t<Rng>>>;
 
+            join_view() = default;
+            explicit join_view(Rng rng)
+              : outer_{view::all(std::move(rng))}
+              // Delay initialization of the other members until begin() is first called,
+              // so that we can create a join_view over e.g. an infinite range of empty ranges,
+              // and avoid hanging if begin is never called.
```

Comment:
I think we need to at least value-initialize the members. Otherwise, they might be singular, and copying/moving the `join_view` is UB, IIUC.